### Machine Learning intern assignment -- HW Saver LLP

In [101]:
#required libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
#import raw data
raw_data = pd.read_csv("Raw_Skills_Dataset.csv") 
raw_data

RAW DATA
0                                           What ifs
1                                          seniority
2                                        familiarity
3                                    functionalities
4                                            Lambdas
...                                              ...
34111                                    negotiation
34112                                      deadlines
34113  Self-motivated, enthusiastic and strong drive
34114                                    negotiation
34115                                      deadlines

[34116 rows x 1 columns]

In [3]:
#import sample data
sample_data = pd.read_csv("Example_Technical_Skills.csv")
sample_data

Technology Skills
0                      SAP Fiori Developer
1    Oracle Instance Management & Strategy
2             Boomi Master Data Management
3    Digital Manufacturing on Cloud ( DMC)
4                                   DevOps
..                                     ...
974        Oracle Cloud Revenue Management
975       Oracle EBS Grid Contral Mgt Pack
976         Amazon Elastic MapReduce (EMR)
977                            Apache Kudu
978                             Oracle ESB

[979 rows x 1 columns]

### Exploratory Data Analysis for Raw data

In [4]:
#information about data
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34116 entries, 0 to 34115
Data columns (total 1 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   RAW DATA  34116 non-null  object
dtypes: object(1)
memory usage: 266.7+ KB


In [5]:
#null values
raw_data.isnull().sum()

RAW DATA    0
dtype: int64

In [6]:
print("No. of data points in data are",raw_data.size)

No. of data points in data are 34116


In [7]:
#shape of dataset
print("No. of Rows and columns in data are",raw_data.shape)

No. of Rows and columns in data are (34116, 1)


In [8]:
#Total count of duplicate values
raw_data.duplicated().sum()

18439

In [9]:
#duplicate words
repeat_words = raw_data.loc[raw_data.duplicated(),:]
repeat_words

RAW DATA
13                                         Hibernate
58                                               SQL
102                                      consistency
114                                              AWS
124           experience building cloud environments
...                                              ...
34107                            limited supervision
34112                                      deadlines
34113  Self-motivated, enthusiastic and strong drive
34114                                    negotiation
34115                                      deadlines

[18439 rows x 1 columns]

In [10]:
#count of duplicate words
repeat_words.value_counts()

RAW DATA         
Python               271
AWS                  238
JavaScript           219
Kubernetes           179
Familiarity          171
                    ... 
attributes             1
audio propagation      1
auditing               1
automated build        1
 AWS experience        1
Length: 3322, dtype: int64

Here Python repeated 271 times and AWS experience repeated once(1)

### Using Machine Learning Models for prediction 

In [21]:
#required libraries
import re
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\P.Shankar\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [13]:
#For raw/test data
raw = []
for i in range(0, 34116):
    cell = re.sub('[^a-zA-Z]', ' ', raw_data['RAW DATA'][i])
    cell = cell.lower()
    cell = cell.split()
    ps = PorterStemmer()
    all_stopwords = stopwords.words('english')
    all_stopwords.remove('not')
    cell = [ps.stem(word) for word in cell if not word in set(all_stopwords)]
    cell = ' '.join(cell)
    raw.append(cell)

In [14]:
#For Train Data
train_set = []
for i in range(0, 979):
    cell = re.sub('[^a-zA-Z]', ' ', sample_data['Technology Skills'][i])
    cell = cell.lower()
    cell = cell.split()
    ps = PorterStemmer()
    all_stopwords = stopwords.words('english')
    all_stopwords.remove('not')
    cell = [ps.stem(word) for word in cell if not word in set(all_stopwords)]
    cell = ' '.join(cell)
    train_set.append(cell)

In [15]:
#using vectorizer to convert to machine understandable language
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 1500)
X = cv.fit_transform(train_set).toarray()
X_test = cv.transform(raw).toarray()
y = np.ones((979,))

#### Gaussian Naive Bayes Model

In [16]:
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X, y)

GaussianNB()

In [27]:
#Predicting on Raw Data

y_pred = classifier.predict(X_test)
prediction1 = []
for i in range(len(X_test)):
    if y_pred[i] == 1:
        prediction1.append(X_test[i])

In [29]:
#Converting the cv transformed data back to Readable Form

prediction1 = cv.inverse_transform(prediction1)
type(prediction1)

list

In [30]:
Final_Words = []
for i in prediction1:
    if list(i) != []:
        Final_Words.append(list(i))
Final_Words

[['function'],
 ['lambda'],
 ['java'],
 ['analysi', 'object'],
 ['databas'],
 ['sql'],
 ['code', 'tool'],
 ['git'],
 ['integr'],
 ['deliveri', 'develop', 'process'],
 ['framework'],
 ['transform'],
 ['analyt', 'report'],
 ['data', 'model'],
 ['standard'],
 ['code'],
 ['optim', 'solut'],
 ['access', 'data', 'servic'],
 ['data'],
 ['data', 'design', 'model', 'physic'],
 ['analyt', 'report', 'solut'],
 ['databas', 'physic'],
 ['integr'],
 ['experi'],
 ['data', 'model', 'physic'],
 ['architectur', 'data', 'enterpris'],
 ['data'],
 ['data', 'warehous'],
 ['data', 'mart'],
 ['model', 'tool'],
 ['tool'],
 ['master'],
 ['code'],
 ['java'],
 ['experi', 'program'],
 ['sql'],
 ['distribut', 'technolog'],
 ['store'],
 ['mysql'],
 ['oracl'],
 ['nosql', 'store'],
 ['big', 'queri'],
 ['distribut', 'engin', 'process'],
 ['apach', 'spark'],
 ['apach'],
 ['distribut'],
 ['apach'],
 ['aw'],
 ['gcp'],
 ['standard'],
 ['gcp'],
 ['azur'],
 ['aw'],
 ['cloud', 'platform', 'technolog'],
 ['cloud', 'experi'],
 

#### K Nearest Neighbor Classifier

In [35]:
from sklearn.neighbors import KNeighborsClassifier
KNN = KNeighborsClassifier()
KNN.fit(X,y)
y_pred = KNN.predict(X_test)

In [36]:
#Predicting on Raw Data
prediction2 = []
for i in range(len(X_test)):
    if y_pred[i] == 1:
        prediction2.append(X_test[i])
        
#Converting the cv transformed data back to Readable Form
prediction2 = cv.inverse_transform(prediction2)
type(prediction2)

Final_Words = []
for i in prediction2:
    if list(i) != []:
        Final_Words.append(list(i))
Final_Words

[['function'],
 ['lambda'],
 ['java'],
 ['analysi', 'object'],
 ['databas'],
 ['sql'],
 ['code', 'tool'],
 ['git'],
 ['integr'],
 ['deliveri', 'develop', 'process'],
 ['framework'],
 ['transform'],
 ['analyt', 'report'],
 ['data', 'model'],
 ['standard'],
 ['code'],
 ['optim', 'solut'],
 ['access', 'data', 'servic'],
 ['data'],
 ['data', 'design', 'model', 'physic'],
 ['analyt', 'report', 'solut'],
 ['databas', 'physic'],
 ['integr'],
 ['experi'],
 ['data', 'model', 'physic'],
 ['architectur', 'data', 'enterpris'],
 ['data'],
 ['data', 'warehous'],
 ['data', 'mart'],
 ['model', 'tool'],
 ['tool'],
 ['master'],
 ['code'],
 ['java'],
 ['experi', 'program'],
 ['sql'],
 ['distribut', 'technolog'],
 ['store'],
 ['mysql'],
 ['oracl'],
 ['nosql', 'store'],
 ['big', 'queri'],
 ['distribut', 'engin', 'process'],
 ['apach', 'spark'],
 ['apach'],
 ['distribut'],
 ['apach'],
 ['aw'],
 ['gcp'],
 ['standard'],
 ['gcp'],
 ['azur'],
 ['aw'],
 ['cloud', 'platform', 'technolog'],
 ['cloud', 'experi'],
 

#### Decision Tree Classifier

In [40]:
from sklearn.tree import DecisionTreeClassifier

In [41]:
DTree_clf = DecisionTreeClassifier(class_weight='balanced')
DTree_clf.fit(X,y)
y_pred = DTree_clf.predict(X_test)

#Predicting on Raw Data
prediction3 = []
for i in range(len(X_test)):
    if y_pred[i] == 1:
        prediction3.append(X_test[i])
        
#Converting the cv transformed data back to Readable Form
prediction3 = cv.inverse_transform(prediction3)
type(prediction3)

Final_Words = []
for i in prediction3:
    if list(i) != []:
        Final_Words.append(list(i))
Final_Words

[['function'],
 ['lambda'],
 ['java'],
 ['analysi', 'object'],
 ['databas'],
 ['sql'],
 ['code', 'tool'],
 ['git'],
 ['integr'],
 ['deliveri', 'develop', 'process'],
 ['framework'],
 ['transform'],
 ['analyt', 'report'],
 ['data', 'model'],
 ['standard'],
 ['code'],
 ['optim', 'solut'],
 ['access', 'data', 'servic'],
 ['data'],
 ['data', 'design', 'model', 'physic'],
 ['analyt', 'report', 'solut'],
 ['databas', 'physic'],
 ['integr'],
 ['experi'],
 ['data', 'model', 'physic'],
 ['architectur', 'data', 'enterpris'],
 ['data'],
 ['data', 'warehous'],
 ['data', 'mart'],
 ['model', 'tool'],
 ['tool'],
 ['master'],
 ['code'],
 ['java'],
 ['experi', 'program'],
 ['sql'],
 ['distribut', 'technolog'],
 ['store'],
 ['mysql'],
 ['oracl'],
 ['nosql', 'store'],
 ['big', 'queri'],
 ['distribut', 'engin', 'process'],
 ['apach', 'spark'],
 ['apach'],
 ['distribut'],
 ['apach'],
 ['aw'],
 ['gcp'],
 ['standard'],
 ['gcp'],
 ['azur'],
 ['aw'],
 ['cloud', 'platform', 'technolog'],
 ['cloud', 'experi'],
 

#### Random Forest Classifier

In [42]:
rt_clf =   RandomForestClassifier()
rt_clf.fit(X,y)
y_pred = rt_clf.predict(X_test)

#Predicting on Raw Data
prediction4 = []
for i in range(len(X_test)):
    if y_pred[i] == 1:
        prediction4.append(X_test[i])
        
#Converting the cv transformed data back to Readable Form
prediction4 = cv.inverse_transform(prediction4)
type(prediction4)

Final_Words = []
for i in prediction4:
    if list(i) != []:
        Final_Words.append(list(i))
Final_Words

<IPython.core.display.Javascript object>

[['function'],
 ['lambda'],
 ['java'],
 ['analysi', 'object'],
 ['databas'],
 ['sql'],
 ['code', 'tool'],
 ['git'],
 ['integr'],
 ['deliveri', 'develop', 'process'],
 ['framework'],
 ['transform'],
 ['analyt', 'report'],
 ['data', 'model'],
 ['standard'],
 ['code'],
 ['optim', 'solut'],
 ['access', 'data', 'servic'],
 ['data'],
 ['data', 'design', 'model', 'physic'],
 ['analyt', 'report', 'solut'],
 ['databas', 'physic'],
 ['integr'],
 ['experi'],
 ['data', 'model', 'physic'],
 ['architectur', 'data', 'enterpris'],
 ['data'],
 ['data', 'warehous'],
 ['data', 'mart'],
 ['model', 'tool'],
 ['tool'],
 ['master'],
 ['code'],
 ['java'],
 ['experi', 'program'],
 ['sql'],
 ['distribut', 'technolog'],
 ['store'],
 ['mysql'],
 ['oracl'],
 ['nosql', 'store'],
 ['big', 'queri'],
 ['distribut', 'engin', 'process'],
 ['apach', 'spark'],
 ['apach'],
 ['distribut'],
 ['apach'],
 ['aw'],
 ['gcp'],
 ['standard'],
 ['gcp'],
 ['azur'],
 ['aw'],
 ['cloud', 'platform', 'technolog'],
 ['cloud', 'experi'],
 

#### Using NLTK models

#### Combining all text cells into single text format

In [43]:
content = raw_data["RAW DATA"]
content

0                                             What ifs
1                                            seniority
2                                          familiarity
3                                      functionalities
4                                              Lambdas
                             ...                      
34111                                      negotiation
34112                                        deadlines
34113    Self-motivated, enthusiastic and strong drive
34114                                      negotiation
34115                                        deadlines
Name: RAW DATA, Length: 34116, dtype: object

In [52]:
n = len(content)
para = []
for i in range(n):
    para.append(content[i])
print(para)

['What ifs', 'seniority', 'familiarity', 'functionalities', 'Lambdas', 'Java Streams', 'Object Oriented analysis', 'Relational Databases', 'SQL', 'ORM', 'JPA2', 'Hibernate', 'MyBatis', 'Hibernate', 'code versioning tools', 'Git.. Familiarity', 'Maven', 'Gradle.. Familiarity', 'continuous integration', 'continuous delivery development processes', 'jenkins', 'bamboo', 'Familiarity', 'frameworks', 'transformations', 'analytical reporting', 'conceptual data models', 'standard naming conventions', 'coding practices', 'consistency', 'optimized solutions', 'methodologies', 'high throughput data access services', 'analyze/profile data', 'conceptual, logical & physical data model designs', 'reporting and analytics solutions', 'physical databases', 'variances', 'discrepancies', 'all phases', 'integrations', 'Demonstrable experience', 'logical and physical data models', 'interpersonal skills', 'modern enterprise data architectures', 'data toolsets', 'e.g.: data warehouse', 'data marts', 'modeling

In [53]:
#Joining list of strings into one
paras = ' '.join(para)
print(paras)

What ifs seniority familiarity functionalities Lambdas Java Streams Object Oriented analysis Relational Databases SQL ORM JPA2 Hibernate MyBatis Hibernate code versioning tools Git.. Familiarity Maven Gradle.. Familiarity continuous integration continuous delivery development processes jenkins bamboo Familiarity frameworks transformations analytical reporting conceptual data models standard naming conventions coding practices consistency optimized solutions methodologies high throughput data access services analyze/profile data conceptual, logical & physical data model designs reporting and analytics solutions physical databases variances discrepancies all phases integrations Demonstrable experience logical and physical data models interpersonal skills modern enterprise data architectures data toolsets e.g.: data warehouse data marts modeling tools profiling tools Bachelor or Master degree STEM majors Strong algorithms coding background either Java Python Scala programming experience E

### Using TF-IDF 

In [54]:
#Import Packages
from nltk import tokenize
from operator import itemgetter
import math

#Declare Variables
doc = paras

#Remove stopwords
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize 
stop_words = set(stopwords.words('english'))

#Find total words in the document 
total_words = doc.split()
total_word_length = len(total_words)
print(total_word_length)

#Find the total number of sentences
total_sentences = tokenize.sent_tokenize(doc)
total_sent_len = len(total_sentences)
print(total_sent_len)

72058
364


In [55]:
#Calculate TF for each word
tf_score = {}
for each_word in total_words:
    each_word = each_word.replace('.','')
    if each_word not in stop_words:
        if each_word in tf_score:
            tf_score[each_word] += 1
        else:
            tf_score[each_word] = 1

# Dividing by total_word_length for each dictionary element
tf_score.update((x, y/int(total_word_length)) for x, y in tf_score.items())
#print(tf_score)

In [56]:
#Function to check if the word is present in a sentence list
def check_sent(word, sentences): 
    final = [all([w in x for w in word]) for x in sentences] 
    sent_len = [sentences[i] for i in range(0, len(final)) if final[i]]
    return int(len(sent_len))

#Calculate IDF for each word
idf_score = {}
for each_word in total_words:
    each_word = each_word.replace('.','')
    if each_word not in stop_words:
        if each_word in idf_score:
            idf_score[each_word] = check_sent(each_word, total_sentences)
        else:
            idf_score[each_word] = 1

# Performing a log and divide
idf_score.update((x, math.log(int(total_sent_len)/y)) for x, y in idf_score.items())
#print(idf_score)

#Calculate TF * IDF
tf_idf_score = {key: tf_score[key] * idf_score.get(key, 0) for key in tf_score.keys()}
#print(tf_idf_score)


In [59]:
#Create a function to get N important words in the document
def get_top_n(dict_elem, n):
    result = dict(sorted(dict_elem.items(), key = itemgetter(1), reverse = True)[:n])
    return result

#Get the top 5 words of significance
print(get_top_n(tf_idf_score, 20))

{'SQL': 0.003138216967072735, '●': 0.002966941772649625, 'Azure': 0.0027675305052025467, 'JavaScript': 0.00258748379584769, 'AWS': 0.002581559477584555, '&': 0.0021411045546030558, '·': 0.002128089699919084, 'development': 0.002018582960237547, 'UI': 0.0020005535061103975, 'Kubernetes': 0.0019748977555391852, 'Familiarity': 0.0018705102636846075, 'skills': 0.0018168825596423747, '•': 0.0016690899607208506, 'experience': 0.0016420978000338403, 'Bachelor’s': 0.0016258125577977727, "Bachelor's": 0.0015340725311156047, 'software': 0.001470658383330835, 'API': 0.0014356165952402954, 'Java': 0.001390147187549483, 'DevOps': 0.0013872656281154047}


In [87]:
skills_data = get_top_n(tf_idf_score, n)
#skills_data = [skills_data] 
skills_data

{'SQL': 0.003138216967072735,
 '●': 0.002966941772649625,
 'Azure': 0.0027675305052025467,
 'JavaScript': 0.00258748379584769,
 'AWS': 0.002581559477584555,
 '&': 0.0021411045546030558,
 '·': 0.002128089699919084,
 'development': 0.002018582960237547,
 'UI': 0.0020005535061103975,
 'Kubernetes': 0.0019748977555391852,
 'Familiarity': 0.0018705102636846075,
 'skills': 0.0018168825596423747,
 '•': 0.0016690899607208506,
 'experience': 0.0016420978000338403,
 'Bachelor’s': 0.0016258125577977727,
 "Bachelor's": 0.0015340725311156047,
 'software': 0.001470658383330835,
 'API': 0.0014356165952402954,
 'Java': 0.001390147187549483,
 'DevOps': 0.0013872656281154047,
 'Python': 0.0013757219977013708,
 'frameworks': 0.0013749698116408099,
 'APIs': 0.001331299692578378,
 'CI/CD': 0.001253218986585755,
 'HTML5': 0.0012462815732345056,
 '-': 0.0011768028230796314,
 'HTML': 0.0011358324482557013,
 'Git': 0.0011356761392699493,
 'Excellent': 0.0010954629634919205,
 '#': 0.0010910057374596415,
 '/': 0

In [99]:
skills = pd.DataFrame(skills_data.items(),columns=['Technical skills', 'TFTDF']) 
skills = skills.drop(["TFTDF"],axis=1)

In [100]:
skills

Technical skills
0                 SQL
1                   ●
2               Azure
3          JavaScript
4                 AWS
...               ...
9780             rest
9781             root
9782           stress
9783               et
9784                 

[9785 rows x 1 columns]

.